File: Java/binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 *
 * Runtime: 0 ms
 * Memory Usage: 37.9 MB	
 *
 */

class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ls = new ArrayList<Integer>();
        preorder(root, ls);
        return ls;
    }
    public void preorder(TreeNode root, List<Integer> ls){
	/* Preorder traversal is visiting : root then left subtree then right subtree */

	/* if node is null, then return back to the caller function */
        if(root==null)
            return;

	/* Add the current node's value to the list */
        ls.add(root.val);

	/* Recurse for left subtree */
        preorder(root.left, ls);

	/* Recurse for right subtree */
        preorder(root.right, ls);
    }
}

File: C++/Binary-Tree-Preorder-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//This solution is an interative solution for problem #144 using stack
//Preorder Traversal is root, left, right
//The Space and Time Complexity is O(n)
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;

        //if root == nullptr, return empty vector
        if (root == NULL) {
            return result;
        }
        
        stack<TreeNode*> node_stack;
        node_stack.push(root);
        
        while (!node_stack.empty()) {
            TreeNode* node = node_stack.top(); 
            node_stack.pop(); //pop the top node from the stack
            result.push_back(node->val); //add the value to the result vector
            
            if (node->right) {
                node_stack.push(node->right); //push right child to the stack if it exists
            }
            if (node->left) {
                node_stack.push(node->left); //push left child to the stack if it exists
            }
        }
        
        return result;
    }
};
